### PR TITLE
Provide a patched Mitsuba kernel pypi package suitable for Eradiate [1/2]

### PR DESCRIPTION
**Pull request by wint3ria**
## Description

Eradiate uses a set of specific Mitsuba variants and implements a few Mitsuba plugins for its functionalities. As a result, the default mitsuba pypi package provided by the RGL is not suitable for the Eradiate workflow. The Eradiate team has thus decided to release both an Eradiate and an [Eradiate Mitsuba](https://pypi.org/project/eradiate-mitsuba/#description) pypi packages. This PR aims at implementing the latter, which includes:
- A Mitsuba kernel with suitable variants compiled for Eradiate
- The Eradiate specific plugins builtin the binary
- A versioning system designed to distinguish the upstream Mitsuba version from the package version and the patch version.

### Compilation configuration

The Eradiate compilation configuration is provided through a `CMakePresets.json` file at the root. The preset is called `eradiate`.
The preset sets the `MI_DEFAULT_VARIANTS` variable to select the variants:
- `scalar_mono`
- `scalar_mono_double`
- `scalar_rgb`
- `scalar_spectral`
- `llvm_rgb`

It also disables [embree](https://www.embree.org/) support due to the [inability of the library to support double precision computations](https://github.com/embree/embree/issues/45#issuecomment-1006053108), necessary for Eradiate. Thus the variable `MI_ENABLE_EMBREE` is set to `OFF`.

### Eradiate plugins

Eradiate plugins have been added from the main Eradiate repository to the `src/eradiate_plugins` directory. The CMake configuration has been updated to include these plugins accordingly:
- `bilambertian`
- `mqdiffuse`
- `rpv`
- `tabphase_irregular`
- `mradiancemeter`
- `hdistant`
- `distantflux`
- `mdistant`
- `sphericalcoordsvolume`

A slight modification has been made to the `sphericalcoordsvolume` plugin, involving renaming a generated using statement manually instead of generating it using [`MI_IMPORT_TYPES`](https://mitsuba.readthedocs.io/en/stable/src/developer_guide/writing_plugin.html#mi-import-types), in order to support the compilation with GCC. The problem here is related to how GCC [(doesn't) work](https://timsong-cpp.github.io/cppwp/n4861/basic.scope.class#2), and is not related to any specific issue with the `MI_IMPORT_TYPES` preprocessing macro.

The following code in `src/eradiate_plugins/volumes/sphericalcoords.cpp:83`

```[c++]
    MI_IMPORT_BASE(Volume, m_to_local, m_bbox)
    MI_IMPORT_TYPES(Volume, VolumeGrid)
```
Has thus been changed to
```[c++]
    MI_IMPORT_BASE(Volume, m_to_local, m_bbox)
    MI_IMPORT_TYPES(VolumeGrid)

    using VolumeType = Volume<Float, Spectrum>;
```
And any subsequent usage of the `Volume` alias has been changed to `VolumeType`.

### Setuptools

The `setup.py` file has been modified in order to pass the package name, package version and the preset name to CI build wheels.

### Versioning

It has been decided to decouple the Mitsuba version from the Eradiate kernel version and the package version.

The Mitsuba version is used in the code and support/disables certain features depending on the hardcoded version in `include/mitsuba/mitsuba.h`. As a result, it is unsuitable for the Eradiate team to work using this version. Not to mention obvious clashes that would occur at any new release of Mitsuba. The current Mitsuba version is 3.2.1 and should only be changed when pulling from the upstream repository.

The patch version for Eradiate has been added to this header. This might be used to identify the Mitsuba binaries potentially provided outside of tree from a pypi package or compiled by users or developers. It shall be used by the Eradiate team to version any subsequent patch to add to the Mitsuba codebase. The Eradiate Mitsuba version is thus provided as follow in `include/mitsuba/mitsuba.h`:

```[c++]
#define ERD_MI_VERSION_MAJOR 0
#define ERD_MI_VERSION_MINOR 0
#define ERD_MI_VERSION_PATCH 1

/// Current release of the Eradiate patch for Mitsuba
#define ERD_MI_VERSION                                                             \
    MI_TOSTRING(ERD_MI_VERSION_MAJOR) "."                                          \
    MI_TOSTRING(ERD_MI_VERSION_MINOR) "."                                          \
    MI_TOSTRING(ERD_MI_VERSION_PATCH)
```

Writing the patch version in the same header limits the code to maintain in case of conflicts, and it "makes sense" to fix conflicts on the version number when releasing a new version or upgrading the upstream.

The package version may differ from the patch version due to release candidates or post release patches. This version is provided by the environment variable `MI_PKG_VERSION` to CI build wheels, but defaults to the Mitsuba version (3.2.1), thus maintaining compatibility if needed.

As a result, this setup allows to :
- check the original mitsuba version, either by looking at the header or in python checking `mitsuba.__version__`
- check the Eradiate patch version by looking at the header
- check the package version in python using `importlib.metadata.version("eradiate_mitsuba")`
- detect a pypi "production" installation, an intree development installation or a custom installation

### CI/CD

The CI/CD developed using Github actions has been slightly adapted:
- Eradiate does not support MS windows, windows support in the eradiate_mitsuba package has been dropped.
- The wheels action passes the github ref in the `MI_PKG_VERSION` to automatically use a version tag.
- An additional input parameter has been added to the wheels action to overwrite the version number in case it is not run on a tag

## Note: documentation

This PR is the first out of 2 patches for the Eradiate Mitsuba engine. The second PR in this regard will focus on the documentation of the features described here and shall provide:
- An updated description and long description of the wheels for pypi
- Documentation of the versioning schema
- Documentation of the build process
- Documentation of the picked variants
- Documentation of the plugins

## Testing

Compilation logs [here](https://github.com/wint3ria/mitsuba3/actions/runs/4882621315)

```[shell]
miskn@0bb02cf8a5d3:~$ pip install eradiate_mitsuba
Collecting eradiate_mitsuba
  Downloading eradiate_mitsuba-0.0.1rc10-cp38-cp38-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (24.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.8/24.8 MB 9.2 MB/s eta 0:00:00
Collecting drjit==0.4.1
  Using cached drjit-0.4.1-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.5 MB)
Installing collected packages: drjit, eradiate_mitsuba
Successfully installed drjit-0.4.1 eradiate_mitsuba-0.0.1rc10
```

TODO: Adapt the Eradiate codebase to support the `eradiate_mitsuba` package, ~and complete the eradiate test cases~

## Checklist

- [x] My code follows the [style guidelines](https://mitsuba.readthedocs.io/en/latest/src/developer_guide.html#coding-style) of this project
- [x] My changes generate no new warnings
- [x] My code also compiles for ~`cuda_*` and~ `llvm_*` variants. If you can't test this, please leave below
- [x] I have commented my code
- [ ] ~I have made corresponding changes to the documentation~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I cleaned the commit history and removed any "Merge" commits
- [x] I give permission that the Mitsuba 3 project may redistribute my contributions under the terms of its [license](https://github.com/mitsuba-renderer/mitsuba/blob/master/LICENSE)

### CODE REVIEW

**Comment by wint3ria** [comment]
I think I'm ready for a first round of review @leroyvn . I don't think we should merge this with the main yet, I would like to implement the support for this in Eradiate to have a better validation of all the functionalities

File: CMakePresets.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+            "CMAKE_BUILD_TYPE": "Release",
+            "MI_ENABLE_EMBREE": "OFF",
+            "MI_DEFAULT_VARIANTS": "scalar_mono;scalar_mono_double;scalar_rgb;scalar_spectral;llvm_rgb"
+        }
+    }],
+    "buildPresets": [{
+        "name": "eradiate",
+        "configurePreset": "eradiate"
+    }]
+}
```
Comment (leroyvn) [inline CMakePresets.json:18]: Newline at EOF:
```suggestion
}

```

File: src/eradiate_plugins/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+message(STATUS "Building Eradiate plugins")
+
+# Plugins
+add_subdirectory(bsdfs)
+add_subdirectory(phase)
+add_subdirectory(sensors)
+add_subdirectory(volumes)
+
+set(MI_PLUGIN_TARGETS "${MI_PLUGIN_TARGETS}" PARENT_SCOPE)
```
Comment (leroyvn) [inline src/eradiate_plugins/CMakeLists.txt:9]:
```suggestion
set(MI_PLUGIN_TARGETS "${MI_PLUGIN_TARGETS}" PARENT_SCOPE)

```

**Review by leroyvn — COMMENTED** [review]
Hi @wint3ria, thanks for this detailed PR. Since we agreed on most points, we're pretty much ready to go with this. Just a question: you write

> The wheels action passes the github ref in the MI_PKG_VERSION to automatically use a version tag.

Doesn't this contradict the in-code version sourcing we eventually decided to go for?

Another thing: are you planning to move plugin tests here as part of this PR or another one?

**Comment by wint3ria** [comment]
> > The wheels action passes the github ref in the MI_PKG_VERSION to automatically use a version tag.
> 
> Doesn't this contradict the in-code version sourcing we eventually decided to go for?
Not really. The idea here is to have access to the patch version in some way. It does not matter if it's inside the header or on the package version.

In practice I don't see a good reason to tag meaningless refs, nor to avoid tagging even a minor release. In practice the header version should be in sync with the tags and the package most of the time. Discrepencies between versions may arise from dev/test/custom environments made by develepers, and can be detected anyway

I do see the use case of being able to test the upload of the package using different versions though. For instance, just for developing the current PR, I made multiple rcs that I eventually dropped. It's just very convenient to be able to override the version when working on the packages themselves.

I made the choice to leave the mitsuba version as detected by setuptools to the mitsuba kernel upstream version. To the best of my knowledge, conditional code involving the version should only be present in the compiled code. So it could be possible to use our own kernel version instead if you prefer so that `import mitsuba; mitsuba.__version__` gives `0.0.1`

**Comment by wint3ria** [comment]
> Another thing: are you planning to move plugin tests here as part of this PR or another one?
Plugin tests? 🙃

**Comment by wint3ria** [comment]
Ok I'll add an additional commit on this PR for the plugin tests.

I'd like to move on with the next feature for the eradiate packaging though, as the schedule is short. Since a package is already available, and the plugin tests are plain python, I can decouple these two tasks and test the plugins in eradiate for now. I already have internal work scheduled right after the 9th of June, thus I may come back to this PR in July, or if I have a hole at some point.

**Comment by wint3ria** [comment]
I took some time for the plugin which tests are now committed :)

**Comment by wint3ria** [comment]
It think this is ready for merging into packaging, a new branch in our mitsuba codebase that we could rebase on the branch upstream/master anytime we pull from the later

File: src/eradiate_plugins/tests/conftest.py
Comment (leroyvn) [inline src/eradiate_plugins/tests/conftest.py:1]: I don't think you need this file at all: the `src` directory already contains all these definitions.

File: src/eradiate_plugins/tests/__init__.py
Comment (leroyvn) [inline src/eradiate_plugins/tests/__init__.py:1]: The Mitsuba test sutie has no `__init__.py`. I suggest to remove them in our code as well. AFAIK the main reason to have them is to minimize the risk of test name clashes: we can simply stick to Mitsuba's test name convention for safety.

**Review by leroyvn — CHANGES_REQUESTED** [review]
Hi @wint3ria, this looks good. Just a few comments and changes and we can merge.

**Pull request merged**